File: src/cpt_anywidget/bhrgt_viewer.py
import pathlib

import anywidget
import traitlets

from cpt_anywidget.vertical import to_vertical

_HERE = pathlib.Path(__file__).parent

# fill for soil names missing from the BRO lithology table
_FALLBACK_COLOR = "#b0b0b0"


def _hex(color):
    r, g, b = (round(c * 255) for c in color)
    return f"#{r:02x}{g:02x}{b:02x}"


def layers_from_bhrgt(bhrgt, vertical_key="depth"):
    """Convert a ``brodata.bhr.GeotechnicalBoreholeResearch`` into the
    ``layers`` trait of :class:`BHRGTViewer`.

    Each descriptive-log layer becomes ``{"top", "bottom", "label",
    "bands"}`` with top/bottom in the requested vertical coordinate
    (``"depth"`` below surface, or a positive-up one like ``"nap"``
    via the borehole's offset).
    Bands are the proportional soil-composition sub-bands from brodata's
    BRO lithology table: ``{"x1", "x2", "color", "hatch"?}`` with x in
    [0, 1] and hatch a matplotlib-style pattern char ("-", "/", "\\\\",
    ".", "o", "|") the front end maps to an SVG pattern.
    """
    from brodata.plot import get_bro_lithology_properties

    table = get_bro_lithology_properties()
    # BRO XML does not guarantee layer order; the front end expects the
    # shallowest layer first
    df = bhrgt.descriptiveBoreholeLog[0]["layer"].sort_values("upperBoundary")

    layers = []
    for row in df.itertuples():
        spec = table.get(row.geotechnicalSoilName)
        if spec is None:
            bands = [{"x1": 0, "x2": 1, "color": _FALLBACK_COLOR}]
        else:
            # base lithologies are a single dict, composites a list of
            # {"width", "color", "hatch"?} sub-bands stacking to <= 1
            if isinstance(spec, dict):
                spec = [{"width": 1, **spec}]
            bands = []
            x = 0.0
            for sub in spec:
                band = {"x1": x, "x2": x + sub["width"], "color": _hex(sub["color"])}
                if "hatch" in sub:
                    band["hatch"] = sub["hatch"]
                x = band["x2"]
                bands.append(band)
        layers.append(
            {
                "top": to_vertical(row.upperBoundary, bhrgt.offset, vertical_key),
                "bottom": to_vertical(row.lowerBoundary, bhrgt.offset, vertical_key),
                "label": row.geotechnicalSoilName,
                "bands": bands,
            }
        )
    return layers


class BHRGTViewer(anywidget.AnyWidget):
    """d3-based geotechnical borehole (BHR-GT) chart: soil-composition
    bands per layer on a shared, zoomable vertical axis, with hover
    readouts and reference-line annotations. Zoom/brush interactions
    match CPTViewer so the two can sit side by side on the same axis.
    """

    _esm = _HERE / "static" / "bhrgt-viewer.js"

    # [{"top", "bottom", "label", "bands": [{"x1", "x2", "color",
    # "hatch"?}, ...]}, ...] — top/bottom in the current vertical
    # coordinate, bands proportional in x [0, 1]; see layers_from_bhrgt
    layers = traitlets.List().tag(sync=True)

    # vertical coordinate the layers are expressed in — a key string
    # ("depth"/"nap" carry display defaults) or a {"key", "label"?,
    # "up"?, "format"?} dict (see vertical.Vertical); only used for the
    # axis label/format — the front end follows layer order
    verticalKey = traitlets.Union(
        [traitlets.Unicode(), traitlets.Dict()], default_value="depth"
    ).tag(sync=True)

    # {verticalKey: [min, max]} override for the vertical axis; the
    # data-driven fallback spans first layer top to last layer bottom
    axisLimits = traitlets.Dict().tag(sync=True)

    # horizontal reference lines (e.g. groundwater): same contract as
    # CPTViewer — {"at", "label", "color"?, "dash"?, "position"?,
    # "offset"?}, "at" in the current vertical coordinate
    annotations = traitlets.List().tag(sync=True)

    # plot size in px; 0 (the default) falls back to the front end's 220x800
    height = traitlets.Int().tag(sync=True)

    width = traitlets.Int().tag(sync=True)

File: js/bhrgt-viewer.ts
import type { RenderProps } from "@anywidget/types";
import { annotationLayer } from "./lib/annotations";
import * as d3 from "./lib/d3";
import { focusRig, haloText } from "./lib/focus-rig";
import { makeVerticalScale, plotClip, verticalAxisTitle, yAxisFor } from "./lib/frame";
import { hatchDefs } from "./lib/hatch";
import type { Annotation, AxisLimits, Band, Layer, VerticalSpec } from "./lib/types";
import { resolveVertical } from "./lib/vertical";
import { verticalZoom } from "./lib/zoom";

/** the synced traits — the TS mirror of BHRGTViewer's traitlets */
interface BhrgtModel {
  layers: Layer[];
  verticalKey: string | VerticalSpec;
  axisLimits: AxisLimits;
  annotations: Annotation[];
  width: number;
  height: number;
}

/** band datum with its layer attached, for placement without lookups */
type LayerBand = Band & { layer: Layer };

export default {
  render({ model, el }: RenderProps<BhrgtModel>) {
    // [{top, bottom, label, bands: [{x1, x2, color, hatch?}]}] — top/bottom
    // in the current vertical coordinate, band x proportional in [0, 1]
    const layers = model.get("layers") ?? [];

    // only the axis label/format depends on this; direction follows the
    // layer order — first layer renders at the top either way
    const vert = resolveVertical(model.get("verticalKey"), "depth");

    // optional {verticalKey: [min, max]} override for the vertical axis
    const axisLimits = model.get("axisLimits") ?? {};

    // horizontal reference lines, e.g. groundwater — same contract as CPT
    const annotations = model.get("annotations") ?? [];

    const width = model.get("width") || 220;
    const height = model.get("height") || 800;

    const marginLeft = 50;
    const marginRight = 20;
    const marginTop = 24;
    const marginBottom = 10;

    const x = d3
      .scaleLinear()
      .domain([0, 1])
      .range([marginLeft, width - marginRight]);

    const first = layers[0];
    const last = layers[layers.length - 1];

    const y = makeVerticalScale(
      [first?.top ?? 0, last?.bottom ?? 1],
      [marginTop, height - marginBottom],
      axisLimits[vert.key],
    );

    const yAxis = yAxisFor(marginLeft, height);

    const svg = d3
      .select(el)
      .append("svg")
      .attr("viewBox", [0, 0, width, height].join(","))
      .attr("width", width)
      .attr("height", height)
      // user-select suppresses text selection during drags/brushes; the
      // -webkit- prefix is still required in Safari
      .attr(
        "style",
        "max-width: 100%; height: auto; user-select: none; -webkit-user-select: none;",
      );

    const clipId = plotClip(svg, "bhrgt-clip", {
      x: marginLeft,
      y: marginTop,
      width: width - marginLeft - marginRight,
      height: height - marginTop - marginBottom,
    });

    // one <pattern> def per hatch char actually used by the data
    const usedHatches = [
      ...new Set(layers.flatMap((l) => (l.bands ?? []).map((b) => b.hatch))),
    ].filter(Boolean) as string[];

    const hatchId = hatchDefs(svg, usedHatches);

    // placed by the zoom drive, like everything else on the vertical scale
    const gy = svg.append("g");

    verticalAxisTitle(svg, vert.label);

    // flatten the layers into one band array up front — band datums carry
    // their layer, so rendering is two flat joins (colour rects + hatch
    // overlays) and placement stays a function of the current zoomed scale
    const bands: LayerBand[] = layers.flatMap((layer) =>
      (layer.bands ?? []).map((b) => ({ ...b, layer })),
    );

    // horizontal extent is static (only y zooms), shared by both joins
    const bandX = (rect: d3.Selection<SVGRectElement, LayerBand, SVGGElement, unknown>) =>
      rect.attr("x", (b) => x(b.x1)).attr("width", (b) => x(b.x2) - x(b.x1));

    const gLayers = svg.append("g").attr("clip-path", `url(#${clipId})`);

    const bandRect = gLayers
      .selectAll<SVGRectElement, LayerBand>("rect")
      .data(bands)
      .join("rect")
      .call(bandX)
      .attr("fill", (b) => b.color)
      .attr("stroke", "white")
      .attr("stroke-width", 0.5);

    const hatchRect = gLayers
      .selectAll<SVGRectElement, LayerBand>(null as unknown as string)
      .data(bands.filter((b) => b.hatch))
      .join("rect")
      .call(bandX)
      .attr("fill", (b) => `url(#${hatchId.get(b.hatch!)})`);

    // white halo keeps names legible over the dark soil bands; visibility
    // is decided per zoom level — labels appear as their layer gets tall
    // enough in pixels
    const soilLabel = gLayers
      .selectAll<SVGTextElement, Layer>("text")
      .data(layers)
      .join("text")
      .attr("x", x(0.5))
      .attr("dy", "0.32em")
      .attr("text-anchor", "middle")
      .attr("font-size", 10)
      .attr("fill", "#333")
      .attr("stroke", "white")
      .attr("stroke-width", 1.5)
      .attr("paint-order", "stroke")
      .text((l) => l.label ?? "");

    const placeLayers = (y1: d3.ScaleLinear<number, number>) => {
      for (const rect of [bandRect, hatchRect]) {
        rect
          .attr("y", (b) => Math.min(y1(b.layer.top), y1(b.layer.bottom)))
          .attr("height", (b) => Math.abs(y1(b.layer.bottom) - y1(b.layer.top)));
      }

      soilLabel
        .attr("y", (l) => (y1(l.top) + y1(l.bottom)) / 2)
        .attr("display", (l) => (Math.abs(y1(l.bottom) - y1(l.top)) >= 14 ? null : "none"));
    };

    const placeAnnotations = annotationLayer(svg, annotations, {
      clipId,
      marginLeft,
      marginRight,
      width,
    });

    // format from the resolved spec: signed for NAP so values near the
    // datum read unambiguously
    const formatVertical = d3.format(vert.format);

    // crosshair at the hovered elevation: the shared skin from the rig,
    // plus this log's own extra readout — the hovered layer's soil name
    const rig = focusRig(svg, { marginLeft, ruleX2: width - marginRight });

    const layerReadout = rig.focus
      .append("text")
      .attr("x", width - marginRight - 6)
      .attr("y", -6)
      .attr("text-anchor", "end")
      .attr("font-size", 12)
      .attr("fill", "#333")
      .call(haloText);

    // bisect matching the layer direction: depth tops ascend, nap tops
    // descend; the candidate is the last layer starting at or above the
    // value, then containment rejects values past its bottom (gaps, ends)
    const bisectTop =
      layers.length < 2 || layers[0].top <= layers[1].top
        ? d3.bisector((l: Layer) => l.top).right
        : d3.bisector((l: Layer, v: number) => v - l.top).right;

    const layerAt = (value: number) => {
      const l = layers[bisectTop(layers, value) - 1];
      return l && value >= Math.min(l.top, l.bottom) && value <= Math.max(l.top, l.bottom)
        ? l
        : undefined;
    };

    function pointermoved(event: PointerEvent) {
      const [, ym] = d3.pointer(event);
      const value = current().invert(ym);
      const layer = layerAt(value);

      if (!layer) {
        rig.hide();
        return;
      }

      rig.show(ym);

      rig.readout.text(`${formatVertical(value)} m`);
      layerReadout.text(layer.label ?? "");
    }

    svg.on("pointerenter pointermove", pointermoved).on("pointerleave", rig.hide);

    // last on purpose: the zoom drive appends the brush overlay, which
    // must stay on top. It runs the initial placement pass and re-places
    // on every zoom
    const current = verticalZoom(svg, {
      y,
      width,
      height,
      marginLeft,
      marginRight,
      marginTop,
      marginBottom,
      placers: [(y1) => gy.call(yAxis, y1), placeLayers, placeAnnotations],
    });
  },
};
